docs/hugo/content/design/ADR-2023-07-Goal-Seeking-KeyVaults/_index.md

```diff
@@ -0,0 +1,154 @@
+
+The [KeyVault](https://azure.github.io/azure-service-operator/reference/keyvault/v1api20210401preview/#keyvault.azure.com/v1api20210401preview.VaultProperties) resource has the following properties relevant to this discussion.
+
+#### createMode
+
+The vault’s create mode to indicate whether the vault need to be recovered or not.
+
+If set to `default` (or if not specified), KeyVault creation will only succeed if there is no existing soft-deleted KeyVault with the same name. If there is such a soft-deleted KeyVault, creation of the KeyVault will fail.
+
+If set to `recover`, KeyVault creation will only succeed if there is an existing soft-deleted KeyVault with the same name. If there is no such KeyVault, creation of the KeyVault will fail.
+
+There is no `purge` option, so it's not possible to purge a KeyVault and create a replacement in a single operation. We can reasonably infer that the Product Group did this deliberately, to prevent accidental purging of a KeyVault.
+
+#### enablePurgeProtection
+
+Specify whether protection against purge is enabled for this vault. 
+
+Setting this property to **true** activates protection against purge for this vault and its content - only the Key Vault service may initiate a hard, irrecoverable deletion. The setting is effective only if soft delete is also enabled. Enabling this functionality is irreversible - that is, the property does not accept **false** as its value.
+  
+#### enableSoftDelete
+
+Specify whether the ‘soft delete’ functionality is enabled for this key vault. 
+
+If it’s not set to any value (**true** or **false**) when creating new key vault, it will be set to true by default. Once set to true, it cannot be reverted to false.
+
+### Option 1: Reinterpretation of createMode
+
+We reinterpret/redefine the permitted values of `createMode` to reflect the goal seeking nature of ASO:
+
+`block` - Create a new KeyVault if there is no existing KeyVault with the same name. If there is an existing KeyVault with the same name, block creation. (No change to current behaviour.)
+
+`recover` - Create a new KeyVault if there is no existing KeyVault with the same name. If there is an existing KeyVault with the same name, recover it. 
+
+The default value would remain be `block`, maintaining the existing behaviour and requiring users to opt-in if they want automatic recovery.
+
+* PRO: Allows users to opt-in to the new behaviour, and to continue to use the existing behaviour if they prefer.
+* PRO: Can be achieved by using our existing supported extension points.
+* CON: Changing the meaning of an existing property might be confusing.
+
+### Option 2: Reterpretation with Purge
+
+As for **Option 1**, but we include an option to purge the KeyVault if it exists.
+
+* PRO: Allows users to configure all possible behaviours.
+* CON: Dangerous, as we could accidentally purge a KeyVault that we weren't intended to purge.
+* CON: The KeyVault Program Group doesn't provide this, likely for good reason.
```

Comment:
>  and a PUT to "purge" would be weird.

I agree that a PUT to purge _this_ KeyVault would be weird, but that's not what I'm writing about. It's about purging _another_ keyVault instead, one that coincidentally has the name we want to use.

To rephrase:

Assume we're doing a PUT to create a new KeyVault.

If we specify `recover` then we're saying "If there's an existing soft-deleted KeyVault with this name, then don't really create a new KeyVault, just recover the old one and I promise to look the other way and won't complain."
This is just a shortcut to avoid the need for a separate PUT to recover the existing KV.

What I'm suggesting is that if we specify `purge` then we're saying "If there's an existing soft-deleted KeyVault with this name, then nuke it before you create a new KeyVault, I promise to look the other way and won't complain."
This would then just be a shortcut to avoid the need for a separate PUT to purge the existing KV.

It would have been easy to build it this way, but the PG didn't - and my guess is they did this because they wanted to make it hard to accidentally delete things. Implementing `createOrPurge` feels like it would be going against that intent.

Reply:
I see what you're saying r.e. "in their context, PUT to purge could have made sense". Fair enough.

I still disagree with your conclusion here though. Even if we assume that the product group considered this option and rejected it because to makes it too easy to accidentally purge (which they may not have, we don't know), that choice makes sense _in their context_ because they have the ability to issue purge as a POST action. So from their raw API perspective saying "they can't do it all in one action, they have to issue the POST first to purge and then they can recreate" isn't problematic.

In _our_ use-case though, it is problematic because we have no such facility to issue POSTs. Everything is resource-driven in ASO so it's not so easy to "just send a post".

I posit that even if the product group intentionally made purge hard, they didn't make it _impossible_ and what we're talking about would be not supporting purge at all. I don't think that makes sense.

We can refer to Terraform for example:
https://github.com/hashicorp/terraform-provider-azurerm/issues/10273
https://github.com/hashicorp/terraform-provider-azurerm/issues/19307

Which shows that the way Terraform "solves" this problem is to allow users to specify the following behaviors:
```
provider "azurerm" {
  features {
    key_vault {
      purge_soft_delete_on_destroy = false
    }
    key_vault_certificate {
      purge_soft_delete_on_destroy = true
    }
    key_vault_key {
      purge_soft_delete_on_destroy = true
    }
    key_vault_secret {
      purge_soft_delete_on_destroy = true
    }
  }
}
```

Which (as their name suggests) will purge immediately after deleting a key, secret, or certificate. The `createOrPurge` (which maybe is a bad name, should be `purgeAndCreate`?) I'm suggesting is significantly safer than that as it's not purging it when it's deleted, instead it's purging it when it's re-used (which means until it's re-used its' still safely soft deleted).

Just as Terraform offers this as an option I think we should too.

Reply:
I think we're on the same wavelength. I'll tweak the wording of the CON.